iwyu.cc

```diff
@@ -3118,42 +3120,68 @@ class InstantiatedTemplateVisitor
     for (const ASTNode* ast_node = current_ast_node();
          ast_node != caller_ast_node_; ast_node = ast_node->parent()) {
       if (ast_node->IsA<TypedefNameDecl>())
-        return Base::VisitSubstTemplateTypeParmType(type);
+        return;
     }
 
     // sizeof(a reference type) is the same as sizeof(underlying type).
     // We have to handle that specially here, or else we'll say the
     // reference is forward-declarable, below.
     if (current_ast_node()->ParentIsA<UnaryExprOrTypeTraitExpr>() &&
-        isa<ReferenceType>(actual_type)) {
-      const ReferenceType* actual_reftype = cast<ReferenceType>(actual_type);
+        isa<ReferenceType>(type)) {
+      const ReferenceType* actual_reftype = cast<ReferenceType>(type);
       ReportTypeUse(CurrentLoc(),
                     actual_reftype->getPointeeTypeAsWritten().getTypePtr());
-      return Base::VisitSubstTemplateTypeParmType(type);
+      return;
     }
 
     // If we're used in a forward-declare context (MyFunc<T>() { T* t; }),
     // or are ourselves a pointer type (MyFunc<Myclass*>()),
     // we don't need to do anything: we're fine being forward-declared.
     if (current_ast_node()->in_forward_declare_context())
-      return Base::VisitSubstTemplateTypeParmType(type);
+      return;
 
     if (current_ast_node()->ParentIsA<PointerType>() ||
         current_ast_node()->ParentIsA<LValueReferenceType>() ||
-        IsPointerOrReferenceAsWritten(actual_type))
-      return Base::VisitSubstTemplateTypeParmType(type);
+        IsPointerOrReferenceAsWritten(type))
+      return;
 
     // We attribute all uses in an instantiated template to the
     // template's caller.
-    ReportTypeUse(caller_loc(), actual_type);
+    ReportTypeUse(caller_loc(), type);
 
     // Also report all previous explicit instantiations (declarations and
     // definitions) as uses of the caller's location.
-    ReportExplicitInstantiations(actual_type);
+    ReportExplicitInstantiations(type);
+  }
+
+  // Our task is made easier since (at least in theory), every time we
+  // instantiate a template type, the instantiation has type
+  // SubstTemplateTypeParmTypeLoc in the AST tree.
+  bool VisitSubstTemplateTypeParmType(clang::SubstTemplateTypeParmType* type) {
+    if (CanIgnoreCurrentASTNode() || CanIgnoreType(type))
+      return true;
+
+    // Figure out how this type was actually written.  clang always
+    // canonicalizes SubstTemplateTypeParmType, losing typedef info, etc.
+    const Type* actual_type = ResugarType(type);
+    CHECK_(actual_type && "If !CanIgnoreType(), we should be resugar-able");
+
+    TemplateTypeParmTypeHelper(actual_type);
 
     return Base::VisitSubstTemplateTypeParmType(type);
   }
 
+  bool VisitTypedefType(clang::TypedefType* type) {
```

Comment:
I have a really hard time with the template visitation. This is the `InstantiatedTemplateVisitor`, right? And I'm guessing it only traverses template instantiations. So why are all typedefs subject to resugaring in instantiated templates?

I'm thinking something like:

``` 
template<class T>
size_t GetTheSize(const T& t) {
  return sizeof(t);
}

int main() {
    float val = 0;
    return (int)GetTheSize(val);
}
```

`size_t` in the `GetTheSize` instantiation is visited here, right?

I wonder if `TemplateTypeParmTypeHelper` should be split into the part that registers the resugar entry, and the lookup? I don't know, it's all very confusing :)

Reply:
Yes, this is `InstantiatedTemplateVisitor`.

The need to track typedefs comes from examples like this:
```c++
template<typename T>
size_t GetTheSize(const T& t) {
  using myT = T;
  return sizeof(myT);
}
```
Without these changes, that template argument would be considered a forward-declare-only use.

Reply:
Right, but I mean that even `size_t` is visited with this logic, even though it doesn't really need to be, right?

Reply:
I don't think `size_t` will be passed to `TemplateTypeParmTypeHelper`.  Which use of `size_t` are you thinking might cause that?

Reply:
I'm really just guessing, but if `InstantiatedTemplateVisitor` traverses instantiated templates, it seems to me a use like:
```
int main() {
    float val = 0;
    return (int)GetTheSize(val);
}
```
Should visit components of `GetTheSize` and find a `typedef` in the return type. But I'll experiment a bit locally and try to bring some facts rather than theoretical worry.

I think my original trigger was that it looked like `TemplateTypeParmTypeHelper` was doing too much (the over-generic name also indicates this), and that maybe it could be split into a populate and lookup pair of methods. But let's hold off on my ideas until I've experimented a little and understand more of what's really going on :)

Reply:
Do you mean because `size_t` is itself a typedef?  It shouldn't cause a call to `TemplateTypeParmTypeHelper` because that typedef will not have been put in the `template_argument_aliases_` map.

I agree the function needs some refactoring.  I want to separate the "is is a full use" part from what is done in response to that, because currently it's a bit inconsistent.  For example, some places therein report a use at `CurrentLoc` and others at `caller_loc`.  I'm pretty sure the former is a mistake.  And only one type of uses leads to a call to `ReportExplicitInstantiations`, which also seems unlikely to be correct.

I also agree that the name isn't great.  Perhaps if I split it up better names will present themselves.